Add Amazon API Gateway OIDC integration guide (#1871)
FusionAuth/fusionauth-site: site/docs/v1/tech/developer-guide/api-gateways/aws-api-gateway.adoc

---
layout: doc
title: Amazon API Gateway
description: Amazon API Gateway OIDC integration
navcategory: developer
---

== Overview

Amazon API Gateway is an AWS service for creating, publishing, maintaining, monitoring, and securing REST, HTTP, and WebSocket APIs at any scale. You can create APIs for AWS services or other web services.

You can configure Amazon API Gateway to handle authentication and authorization to a resource through JSON Web Tokens (JWTs) issued on behalf of a user by an identity provider.

In this document, you'll learn how to set up Amazon API Gateway and FusionAuth as the identity provider to protect a lambda function running on your AWS cloud instance.

== Prerequisites

> * A FusionAuth instance running on a publicly accessible URL. You can spin up a link:/pricing[basic FusionAuth Cloud instance] or link:/docs/v1/tech/installation-guide/[install it on any server].

Reviewer: 👍

Reviewer: Can you add something about exposing a local FusionAuth using ngrok or a similar tool?

Finally, click `Create` on the [breadcrumb]#Review and Create# page.

A summary page for the API should appear. Copy the [field]#Invoke URL# from the [breadcrumb]#$default# stage under the stages section. This will be the base URL for the API. We'll add a lambda function next, which will give us a route to call.

=== Create a lambda function to secure

In the main AWS search bar, search for and select `lambda`.

Click `Create function` to create a new lambda function. Name the function `privateFunction` and leave all the other options as the defaults. Click `Create Function` at the bottom of the page.

image::api-gateways/aws-api-gateway/name-lambda.png[Name the Lambda,width=1200]

Under the [breadcrumb]#Code# tab for the lambda, click on the "index.mjs" file and replace the contents with the following code to return the claims found in the JWT:

[source,javascript,title="Lambda function to reflect user claims"]
----
export const handler = async(event) => {
    const claims = event.requestContext.authorizer.claims;
    const response = {
        statusCode: 200,
        body: JSON.stringify(claims),
    };
    return response;
};
----

> Then click the `Deploy` button to update the lambda.

Author: noted

Click the expander next to [breadcrumb]#Function overview# near the top of the page. Then click on `Add Trigger`.

Under [field]#Trigger Configuration#, choose `API Gateway` as the source. Select `Use an existing API` and then select the API gateway set up earlier. Select the `$default` deployment stage.

Under [field]#Security#, choose `Create a JWT authorizer`.

Set [field]#Identity source# to `$request.header.Authorization`.

> For [field]#Issuer#, set the URL to be the same as you set in FusionAuth for the JWT issuer in the Tenant configuration. Make sure that it is exactly the same, including trailing slashes, without any whitespace.

Reviewer: ```suggestion
For API Gateway [field]#Issuer#, set the value to be the same URL as you set for the [field]#Issuer# in the FusionAuth Tenant configuration. Make sure that it is exactly the same, including trailing slashes, without any whitespace.
```

Under [field]#Audiences#, paste in the `Client ID` from the FusionAuth application created earlier.

Then click the primary `Add` button at the bottom right of the screen.

After completing this, a new route will be added to your gateway API, with the same name as the lambda, `privateFunction` here. Concat this name with the "Invoke URL" recorded earlier when setting up the API gateway to get the full URL. It should look similar in form to this: `\https://w3miabt10d.execute-api.us-east-2.amazonaws.com/privateFunction`.

> If you call this route through a browser, Postman, or cURL, you should receive a 401 and an `Unauthorized` message. We'll create a token next with FusionAuth to show how the route can be successfully called.

Reviewer: ```suggestion
If you call this route through a browser, Postman, or cURL, you will receive a `401` HTTP status code and an `Unauthorized` message. We'll create a token next with FusionAuth to show how the route can be successfully called.
```

> == Testing with a token from FusionAuth

Reviewer: ```suggestion
== Testing With a Token From FusionAuth
```

> To access the secured lambda function, we'll need to get a valid token from FusionAuth. Normally, your frontend app would redirect a user to FusionAuth to login, obtain a code, and then exchange that code for a token, using a framework of your choice. For the purposes of this guide, we'll call the FusionAuth API on behalf of a user to create a token directly.

Reviewer: ```suggestion
To access the secured lambda function, we'll need to get a valid token. Normally, your frontend app would use the Authorization Code grant which would redirect a user to FusionAuth to login, obtain a code, and then exchange that code for a token, using a framework of your choice. link:/docs/v1/tech/oauth/[More on the Authorization Code grant.]

For the purposes of this guide, we'll call the link:/docs/v1/tech/apis/login[FusionAuth Login API] on behalf of a user to create a token directly. This is a more straightforward means of getting a token, but the generated token will be similar with either process.
```

> To enable this, there are a few settings that need to be changed on FusionAuth. In the side panel in FusionAuth, select [breadcrumb]#Settings > API Keys#. Click the green `+` button at the top left to create a new key.

Reviewer: ```suggestion
To use the Login API, there are a few settings that need to be changed on FusionAuth. In the side panel in FusionAuth, select [breadcrumb]#Settings > API Keys#. Click the green `+` button at the top left to create a new key.
```

Give the key a meaningful description, like `API Gateway Key`. Select the Tenant that you created the FusionAuth application under (normally `Default`).

> Then, under [breadcrumb]#Endpoints#, allow `POST` on the `/api/login` route.

Reviewer: ```suggestion
Then, under [breadcrumb]#Endpoints#, allow `POST` on the `/api/login` route. This key will only be allowed to all this endpoint with the `POST` method.
```

Author: *call* this endpoint I guess?

Reviewer: Correct! `This key will only be allowed to call`

image::api-gateways/aws-api-gateway/login-api.png[Enabling the login API,width=1200]

> Save the API key. You should now see a list of all API keys. Click on the lock icon next to the [field]#key# field for the key you just created and copy out the API key for later use.

Reviewer: ```suggestion
Save the API key. You should now see a list of all API keys. Click on the lock icon next to the [field]#key# field for the key you just created and copy the API key for later use. Make sure you get the API key value, and not the API key Id.
```

image::api-gateways/aws-api-gateway/api-key.png[Enabling the login API,width=1200,role=bottom-cropped]

> Now create a test user to login with on FusionAuth. Navigate to [breadcrumb]#Users# in the sidebar and click the green `+` button at the top right of the screen to create a new user. Choose the appropriate tenant (usually `default`), and fill in the information. You can choose not to send an email to set up the password but rather set it directly when creating the user by toggling the [field]#Send email to set up password# switch. Record this user email and password, as we'll use it to obtain a token.

Reviewer: Any reason to set up a test user? Any reason not to use the current user (since they were able to log into FusionAuth?)

After saving the user, click the [breadcrumb]#Registrations# tab. Click `Add registrations` and select the FusionAuth application you set up earlier.

image::api-gateways/aws-api-gateway/registrations-tab.png[The application registrations tab,width=1200,role=bottom-cropped]

> Using cURL or Postman, make a POST request to your FusionAuth instance's `/api/login` endpoint. This will return a JWT in the response.

Reviewer: Let's use curl, not cURL, throughout the doc, please.

The cURL call should look like this:

[source,shell,title="cURL command to obtain a JWT token"]
----
curl --location --request POST 'https://<YOUR_FUSIONAUTH_URL>/api/login' \
--header 'Authorization: <API_KEY>' \
--header 'Content-Type: application/json' \
--data-raw '  {
    "loginId": "<USER_EMAIL>",
    "password": "<USER_PASSWORD>",
    "applicationId": "<APPLICATION_ID>",
    "noJWT" : false
  }'
----

Where:

* `<YOUR_FUSIONAUTH_URL>` is the URL of your FusionAuth installation.
* `<API_KEY>` is the API created above, enabled for login.
* `<USER_EMAIL>` is the email address of the user added above.

> * `<USER_PASSWORD>` is the password of the user added above.

Reviewer: missing <APPLICATION_ID>

Reviewer: @sixhobbits ?

The return response from FusionAuth should look similar to the following:

> [source,json,title="Return token from the API call"]

Reviewer: ```suggestion
[source,json,title="Login API Response"]
```

----
{
    "token": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImhDUjA4X3daR2s0OUFlYUFmRDY5ZmJKWmRGTSJ9.eyJhdWQiOiI2M2I3M2Y3Ni03NDAwLTQ4N2QtYjEyMi01NzA1Yjg0OGRhODAiLCJleHAiOjE2NzMzNjYyMDQsImlhdCI6MTY3MzM2MjYwNCwiaXNzIjoiaHR0cHM6Ly9mdXNpb25hdXRoLnJpdHphLmNvIiwic3ViIjoiMzk2MzAwMGYtNjg2ZC00MTY5LWI2MjgtOWM5YzQ1MzRiNzgwIiwianRpIjoiZDk3ZGIyZWYtZjExNS00ZDIxLWFlOTQtMDIyN2RmMGU4YzI5IiwiYXV0aGVudGljYXRpb25UeXBlIjoiUEFTU1dPUkQiLCJlbWFpbCI6ImJvYkBhd3MuY29tIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsInByZWZlcnJlZF91c2VybmFtZSI6ImJvYmF3cyIsImFwcGxpY2F0aW9uSWQiOiI2M2I3M2Y3Ni03NDAwLTQ4N2QtYjEyMi01NzA1Yjg0OGRhODAiLCJyb2xlcyI6W10sImF1dGhfdGltZSI6MTY3MzM2MjYwNCwidGlkIjoiZjAwNGMxZmUtNDg0Yi05MDJjLWQ3Y2EtYmRiYzQ2NGRhMGI3In0.m7gzXhNLToPNVE1p5Vo2pLgP6WBcPNfS_zZJnJ81mdEgi6-orViz-tU8j0L8wva0-8KlMdy54cq_XjnDnYJ0aX90O4ZE_QVU5NuDDfzXH14wQtKQoIIydsB6ZvQoBt8JNFUHJb9ANLCGnfn6FVQKqPIzye18Gx_7wYSVokw3eLNFyzrq9dwOD5Q8V9gvZmXV2pTokQAtA7qFaadb2dIeFlSEB7wamKiZLXILjeWAeMbbvAAMQZWFh46UJjwr06QTd8PxQmRwDWWznJy1Vs8EAgZA4vkRSWnn3IbiaCtOaL1ANuEex6il7q32ahxj0Ncm9wn0DbDsQE9NB0CCNTSIhA",
    "tokenExpirationInstant": 1673366204805,
    "user": {

>         ....

Reviewer: ```suggestion
```

Reviewer: these look red and angry when syntax highlighed because they aren't valid json. Don't have a real good option here, so omit them.

    }
}
----

> Copy the `token` value out. This is the JWT, which we can use to access the API gateway function.

Reviewer: ```suggestion
Copy the `token` value. This is a JWT, which we can use to access the API gateway function.
```

Now call the API gateway with the following cURL command:

[source,shell,title="Calling the API Gateway endpoint with the JWT"]
----
curl --location --request GET 'https://<GATEWAY_URL>/privateFunction' \
--header 'Authorization: Bearer <JWT_TOKEN>'
----

Where:

* `<GATEWAY_URL>` is the invoke URL saved earlier.
* `<JWT_TOKEN>` is the JWT from the `/api/login` call.

You should see the function return with a reflection of the claims it received:

[source,json,title="Return token from the API call"]
----
{
    "applicationId":"63b73f76-7400-487d-b122-5705b848da80",
    "aud":"63b73f76-7400-487d-b122-5705b848da80",
    "auth_time":"1672137288",
    "authenticationType":"PASSWORD",
    "email":"ehrlich@piedpiper.com",
    "email_verified":"true",
    "exp":"1672140888",
    "iat":"1672137288",
    "iss":"https://local.fusionauth.io",
    "jti":"2f80b975-1870-4970-b876-90c2e7d9444b",
    "preferred_username":"erlich",
    "roles":"[]",
    "sub":"3963000f-686d-4169-b628-9c9c4534b780",
    "tid":"f004c1fe-484b-902c-d7ca-bdbc464da0b7"
 }
----

> 

Reviewer: Should add a troubleshooting section.

This won't work with a local installation, because the API gateway needs to be able to access the JWKS.json to validate the key.

If you don't set up asymmetric keys, it will fail. The key must be an RSA key as well.

If you wait too long between when you run the login API and when you call the API and the token expires it will fail.

== Next Steps

> You can build an entire HTTP API using Amazon API Gateway, lambda, and FusionAuth without having to manage any servers.

Reviewer: ```suggestion
You can build a complete HTTP API using Amazon API Gateway, AWS Lambda, and FusionAuth without managing any servers.
```

> You can also configure required scopes for a route and check against them before allowing a user through.

Reviewer: ```suggestion
You can also configure required scopes for a route and check against them before allowing a user through. Set the `scope` claim using a link:/docs/v1/tech/lambdas/jwt-populate[JWT Populate lambda].
```

> Finally, you can tweak your FusionAuth settings to ensure that the user is registered for the Amazon Gateway API application or fire off webhooks when the user logs in.

Reviewer: ```suggestion
Finally, you can configure FusionAuth to ensure that the user is registered for the Amazon Gateway API application or link:/docs/v1/tech/events-webhooks/[fire off webhooks] when the user logs in.
```
